src/Planner/PlannerJoins.cpp

```diff
@@ -385,6 +600,203 @@ void buildJoinClause(
     }
 }
 
+JoinClauses buildJoinClauses(
+    ActionsDAG & left_dag,
+    ActionsDAG & right_dag,
+    ActionsDAG & joined_dag,
+    const PlannerContextPtr & planner_context,
+    const QueryTreeNodePtr & join_expression,
+    const TableExpressionSet & left_table_expressions,
+    const TableExpressionSet & right_table_expressions,
+    const JoinNode & join_node)
+{
+    if (join_expression->getNodeType() != QueryTreeNodeType::FUNCTION)
+    {
+        JoinClauses result;
+        result.emplace_back();
+        buildSimpleJoinClause(
+            left_dag,
+            right_dag,
+            joined_dag,
+            planner_context,
+            join_expression,
+            left_table_expressions,
+            right_table_expressions,
+            join_node,
+            result.front());
+        return result;
+    }
+
+    std::unordered_map<IQueryTreeNode *, JoinClauses> built_clauses;
+    std::stack<QueryTreeNodePtr> nodes_to_process;
+    nodes_to_process.push(join_expression);
+
+    while (!nodes_to_process.empty())
+    {
+        auto node = nodes_to_process.top();
+        auto * function_node = node->as<FunctionNode>();
+        const auto function_name = function_node ? function_node->getFunctionName() : String();
+        const auto expression_sides
+            = extractJoinTableSidesFromExpression(node.get(), left_table_expressions, right_table_expressions, join_node);
+        // If the expression is a logical expression and it contains expressions from both sides, let's combine the clauses, otherwise let's just build one join clause
+        if ((function_name == "and" || function_name == "or") && expression_sides.size() == 2)
+        {
+            auto & arguments = function_node->getArguments().getNodes();
+            auto * first_argument = arguments.front().get();
+            if (const auto it = built_clauses.find(first_argument); it == built_clauses.end())
+            {
+                for (auto & argument : arguments)
+                    nodes_to_process.push(argument);
+
+                continue;
+            }
+
+            nodes_to_process.pop();
+            JoinClauses result;
+            if (function_name == "or")
+            {
+                for (auto & argument : arguments)
+                {
+                    auto & child_res = built_clauses.at(argument.get());
+                    result.insert(result.end(), std::make_move_iterator(child_res.begin()), std::make_move_iterator(child_res.end()));
```

Comment:
Yes, we shouldn't reach the same node. The key of the map is `IQueryTreeNode *`, so as far as I understood we shouldn't reach the same node again. However I will add that `clear()` call. It doesn't take much but makes the code more robust. Debugging random crashes because of memory violation is not fun.

Reply:
Yes, we shouldn't reach the same node. The key of the map is `IQueryTreeNode *`, so as far as I understood we shouldn't reach the same node again. However I will add that `clear()` call. It doesn't take much but makes the code more robust. Debugging random crashes because of memory violation is not fun.